plugins/modules/units.py

```diff
@@ -0,0 +1,110 @@
+## along with this program.  If not, see <http://www.gnu.org/licenses/>. ##
+##                                                                       ##
+###########################################################################
+
+
+# This module file contains functions to read and convert numbers with units
+
+import re
+
+_ftin_re = re.compile("^(-?)(?:(\\d+(?:\\.\\d+)?)(ft|')) ?(?:(\\d+(?:\\.\\d+)?)(in|\"))$") # Regex for combined feet/inch measures
+_numunit_re = re.compile(r"^(-?\d+(?:\.\d+)?) ?(\D.*)?$") # Regex for any number followed by an optional unit
+_si_prefixes = {
+    "n": 1E-9, "u": 1E-6, "µ": 1E-6, "m": 0.001, "c": 0.01, "d": 0.1, "da": 10, "h": 100, "k": 1000, "M": 1E6, "G": 1E9, "T": 1E12,
+    "nano": 1E-9, "micro": 1E-6, "milli": 0.001, "centi": 0.01, "deci": 0.1, "deca": 10, "hecto": 100, "kilo": 1000, "mega": 1E6, "giga": 1E9, "tera": 1E12,
+}
+
+
+# Converts strings that contain a number with a unit to a dict of {"value": float, "unit": string}
+# Input:
+#   string: the string to parse, e.g. "22.4 km"
+#   defaultUnit: the unit if not specified (default: None)
+# Returns:
+#   None if conversion isn't possible
+#   Otherwise a dict with keys:
+#       value [float] - the number in the string
+#       unit [string or None] - the unit
+# For a string with multiple numbers/units, it converts it to a float of the largest unit (e.g. 3'4" becomes 3.33 ft)
+def parseNumberUnitString(string, defaultUnit = None):
+    if not string:
+        return None
+    string = string.strip()
+
+    # Combined units: feet and inch
+    m = re.fullmatch(_ftin_re, string)
+    if m:
+        return {
+            "value": float(m.group(1) + m.group(2)) + float(m.group(1) + m.group(4))/12,
+            "unit": "ft"
+        }
+    # Regular numbers with optional unit
+    m = re.fullmatch(_numunit_re, string)
+    if m:
+        return {
+            "value": float(m.group(1)),
+            "unit": m.group(2) or defaultUnit
+        }
+    # Not a numerical value
+    return None
+
+
+# Converts a number in arbitrary units to the default unit
+# Input:
+#   x: either a string with a number + optional unit to parse, or a dict with value:[float] and unit:[string] keys
+#   convertTo: the unit to convert the x-input to (the abbreviation)
+#   If x is a string without unit, it assumes the default unit equals the unit of convertTo
+# Returns:
+#   None if the input was None or the string couldn't be parsed into a number + optional unit
+#   The value [float] converted to convertTo units otherwise
+# Throws:
+#   If conversion wasn't possible
+def convertToUnit(x, convertTo):
+    if isinstance(x, str):
+        # String input, extract value and unit
+        x = parseNumberUnitString(x, convertTo)
+    if not x:
+        return None
+    if x["unit"] == convertTo:
+        return x["value"]
+
+    # Length based conversions
+    if convertTo == "m": # default for range, height, length
```

Comment:
Can we use pint once the parsing is done ?

Reply:
It would still require a lot more pre-parsing from our side than I initially expected. Pint for instance doesn't natively support things like `m3` (for the fix of 2251) or `"` (inch, as symbol, for siunit_length). So we'd still need a dictionary mapping for instance `'` to `ft`, `m3` to `m³` etc.

In addition, feeding pint with arguments like `value=123.45, units='m'` gives issues because Pint then complains about the use of scaling factors (it seems to assume it's milli on first glance), so it'd have to become something like `x["value"] * ureg("1" + x["unit"])` to use it (with `ureg` the function from Pint)
Lastly, it still does mathematics if there's anything in the unit. E.g. if the unit contains `30m*3` pint will simply multiply the value by 3 instead of complaining about invalid units.

So personally, after noticing these things I think I would prefer to have a function of our own until for instance Pint has an option to disable math or we need many more conversion functions. There are benefits to each, but...

Reply:
Ok. Thank you for the explanation.